src/bufread.rs

```diff
@@ -186,14 +193,19 @@ impl<R> BzDecoder<R> {
 
 impl<R: BufRead> Read for BzDecoder<R> {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        if self.done {
-            return Ok(0)
-        }
         loop {
-            let (read, consumed, eof, ret);
+            let (read, consumed, ret);
             {
                 let input = try!(self.obj.fill_buf());
-                eof = input.is_empty();
+                if input.is_empty() {
```

Comment:
At least for zlib this wasn't quite right where we need to wait for the `decompress` call below to return EOF before we return EOF to the user.